packages/client/src/vscode/apiWrapper.ts

```diff
@@ -0,0 +1,333 @@
+        const userServices = this.apiConfig.serviceOverrides;
+        const haveThemeService = Object.keys(userServices).includes('themeService');
+        const haveTextmateService = Object.keys(userServices).includes('textMateTokenizationFeature');
+        const haveMarkersService = Object.keys(userServices).includes('markersService');
+        const haveViewsService = Object.keys(userServices).includes('viewsService');
+
+        // theme requires textmate
+        if (haveThemeService && !haveTextmateService) {
+            throw new Error('"theme" service requires "textmate" service. Please add it to the "userServices".');
+        }
+
+        // markers service requires views service
+        if (haveMarkersService && !haveViewsService) {
+            throw new Error('"markers" service requires "views" service. Please add it to the "userServices".');
+        }
+    }
+
+    /**
+     * monaco-vscode-api automatically loads the following services:
+     *  - layout
+     *  - environment
+     *  - extension
+     *  - files
+     *  - quickAccess
+     * monaco-languageclient always adds the following services:
+     *   - languages
+     *   - log
+     *   - model
+     */
+    protected async importAllServices(instructions?: InitServicesInstructions) {
+        const services = await this.supplyRequiredServices();
+
+        mergeServices(services, this.apiConfig.serviceOverrides);
+        await configureExtHostWorker(this.apiConfig.advanced?.enableExtHostWorker === true, services);
+
+        reportServiceLoading(services, this.logger);
+
+        if (instructions?.performServiceConsistencyChecks === true) {
+            this.checkServiceConsistency();
+        }
+
+        if (this.apiConfig.viewsConfig?.viewServiceType === 'ViewsService' || this.apiConfig.viewsConfig?.viewServiceType === 'WorkspaceService') {
+            await initialize(services, this.apiConfig.htmlContainer, this.apiConfig.workspaceConfig, this.apiConfig.envOptions);
+        } else {
+            await initialize(services, undefined, this.apiConfig.workspaceConfig, this.apiConfig.envOptions);
+        }
+
+        setUnexpectedErrorHandler((e) => {
+            this.logger.createErrorAndLog('Unexpected error', e);
+        });
+    }
+
+    async initExtensions(): Promise<void> {
+        if (this.apiConfig.$type === 'extended' && (this.apiConfig.advanced?.loadThemes === undefined ? true : this.apiConfig.advanced.loadThemes === true)) {
+            await import('@codingame/monaco-vscode-theme-defaults-default-extension');
+        }
+
+        const extensions = this.apiConfig.extensions;
+        if (this.apiConfig.extensions) {
+            const allPromises: Array<Promise<void>> = [];
+            const extensionIds: string[] = [];
+            getExtensionManifests().forEach((ext) => {
+                extensionIds.push(ext.identifier.id);
+            });
+            for (const extensionConfig of extensions ?? []) {
```

Comment:
Lots of nesting in here, any chance we can flatten this a bit?

Reply:
Yes, a function treating a single extension should be extracted.